### PR TITLE
OTG: gNMI-1.3: Benchmarking: Drained Configuration Convergence Time

### LINKED ISSUE #1190: OTG: gNMI-1.3: Benchmarking: Drained Configuration Convergence Time

### PR DESCRIPTION
**Pull request by ANISH-GOTTAPU**
OTG conversion of OTG: gNMI-1.3: Benchmarking: Drained Configuration Convergence Time [here](https://github.com/openconfig/featureprofiles/tree/main/feature/experimental/system/gnmi/benchmarking/ate_tests/drained_configuration_convergence_time)

resolves #1190 

Log:
[gNMI_1.3.log](https://github.com/openconfig/featureprofiles/files/14129047/gNMI_1.3.log)(https://drive.google.com/file/d/1y623k7EvwZNHB3lHvbJ5sHneNvf-xGSE/view?usp=share_link)

### CODE REVIEW

**Review by jasdeep-hundal — CHANGES_REQUESTED** [review]

File: feature/experimental/system/gnmi/benchmarking/otg_tests/drained_configuration_convergence_time/drained_configuration_convergence_time_isis_test.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+				metric, present := v.Val()
+				if present {
+					if metric == setup.ISISMetric {
+						return true
+					}
+				}
+				return false
+			}).Await(t)
+
+			metricInReceivedLsp := gnmi.GetAll(t, ate.OTG(), gnmi.OTG().IsisRouter("devIsis"+ap.Name()).LinkStateDatabase().LspsAny().Tlvs().ExtendedIpv4Reachability().PrefixAny().Metric().State())[0]
```
Comment (jasdeep-hundal) [inline feature/experimental/system/gnmi/benchmarking/otg_tests/drained_configuration_convergence_time/drained_configuration_convergence_time_isis_test.go:77]: Can just get the last value from the WatchAll calls first returned value, shouldn't need to use the additional GetAll here.

File: feature/experimental/system/gnmi/benchmarking/otg_tests/internal/setup/setup.go [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+// ConfigureATE function is to configure ate ports with ipv4 , bgp
+// and isis peers.
+func ConfigureATE(t *testing.T, ate *ondatra.ATEDevice) {
+	otg := ate.OTG()
+	topo := otg.NewConfig(t)
+
+	for i, dp := range ate.Ports() {
+		ISISMetricList = append(ISISMetricList, ISISMetric)
+		ISISSetBitList = append(ISISSetBitList, true)
+		atePortAttr := attrs.Attributes{
```
[the remark this answers is not in the document]
Comment (jasdeep-hundal) [reply feature/experimental/system/gnmi/benchmarking/otg_tests/internal/setup/setup.go:271]: IP should just be in the DUTIPList correct? And that should be the only value you need in the attributes object for the DUT interface.

**Review by jasdeep-hundal — CHANGES_REQUESTED** [review]

File: feature/experimental/system/gnmi/benchmarking/otg_tests/drained_configuration_convergence_time/drained_configuration_convergence_time_isis_test.go (hunk 1 above)
Comment (ANISH-GOTTAPU) [reply feature/experimental/system/gnmi/benchmarking/otg_tests/drained_configuration_convergence_time/drained_configuration_convergence_time_isis_test.go]: yeah agree, changed

File: feature/experimental/system/gnmi/benchmarking/otg_tests/internal/setup/setup.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ANISH-GOTTAPU) [reply feature/experimental/system/gnmi/benchmarking/otg_tests/internal/setup/setup.go:271]: Yes, DUTIPList should suffice as the attributes object for the DUT interface. I'm referring to the subsequent step in the same function where the BGP and ISIS configurations are performed

I need to create one more function to retrieve the `dev` object based on port again

```
		// Add BGP routes and ISIS routes , ate port1 is ingress port.
		if dp.ID() == "port1" {
			// Add BGP on ATE
			bgpDut1 := dev.Bgp().SetRouterId(ip.Address())
			bgpDut1Peer := bgpDut1.Ipv4Interfaces().Add().SetIpv4Name(ip.Name()).Peers().Add().SetName(dp.ID() + ".BGP4.peer")
			bgpDut1Peer.SetPeerAddress(DUTIPList[dp.ID()].String()).SetAsNumber(ATEAs2).SetAsType(gosnappi.BgpV4PeerAsType.EBGP)

			bgpDut1Peer.Capability().SetIpv4Unicast(true)
			bgpDut1Peer.LearnedInformationFilter().SetUnicastIpv4Prefix(true)
```

Referring to this line https://github.com/open-traffic-generator/featureprofiles/blob/otg-gnmi1.3/feature/experimental/system/gnmi/benchmarking/otg_tests/internal/setup/setup.go#L291

File: feature/experimental/system/gnmi/benchmarking/otg_tests/drained_configuration_convergence_time/drained_configuration_convergence_time_isis_test.go [hunk 3]
```diff
@@ -74,7 +74,7 @@ func verifyISISMetric(t *testing.T, dut *ondatra.DUTDevice, ate *ondatra.ATEDevi
 				return false
 			}).Await(t)
 
-			metricInReceivedLsp := gnmi.GetAll(t, ate.OTG(), gnmi.OTG().IsisRouter("devIsis"+ap.Name()).LinkStateDatabase().LspsAny().Tlvs().ExtendedIpv4Reachability().PrefixAny().Metric().State())[0]
+			metricInReceivedLsp, _ := got.Val()
```
[the remark this answers is not in the document]
Comment (ANISH-GOTTAPU) [reply feature/experimental/system/gnmi/benchmarking/otg_tests/drained_configuration_convergence_time/drained_configuration_convergence_time_isis_test.go:77]: are you suggesting this way? 
```
			metricInReceivedLsp, ok := gnmi.WatchAll(t, ate.OTG(), gnmi.OTG().IsisRouter("devIsis"+ap.Name()).LinkStateDatabase().LspsAny().Tlvs().ExtendedIpv4Reachability().PrefixAny().Metric().State(), time.Minute, func(v *ygnmi.Value[uint32]) bool {
				metric, present := v.Val()
				if present {
					if metric == setup.ISISMetric {
						return true
					}
				}
				return false
			}).Await(t)

			if !ok {
				t.Fatalf("Metric not matched. Expected %d got %d ", setup.ISISMetric, metricInReceivedLsp.Val())
			}
```

This way **metricInReceivedLsp** returns two values **(int bool)** . Please let me know if I am missing anything.

[the remark this answers is not in the document]
Comment (jasdeep-hundal) [reply feature/experimental/system/gnmi/benchmarking/otg_tests/drained_configuration_convergence_time/drained_configuration_convergence_time_isis_test.go:77]: Ah, my mistake, forgot that Val() returns two values. We can leave this as it is now.

File: feature/experimental/system/gnmi/benchmarking/otg_tests/internal/setup/setup.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jasdeep-hundal) [reply feature/experimental/system/gnmi/benchmarking/otg_tests/internal/setup/setup.go:271]: Ah, I understand now.

In that case, do we need the `atePortAttr` object at all? Can we just use `ATEIPList[dp.ID()]` and `plenIPv4` directly?

**Review by jasdeep-hundal — CHANGES_REQUESTED** [review]
Also needs:
- Code changes from: https://github.com/openconfig/featureprofiles/commit/b0a00495b8028989dfbd88c72200d31ef6b66f7e
- Updates for the new way of referencing deviations.

File: feature/experimental/system/gnmi/benchmarking/otg_tests/internal/setup/setup.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ANISH-GOTTAPU) [reply feature/experimental/system/gnmi/benchmarking/otg_tests/internal/setup/setup.go:274]: Yeah, you are correct `atePortAttr` object is not needed. I followed ATE test so adopted the same, changed it now

**Comment by ANISH-GOTTAPU** [comment]
> Also needs:
> 
> * Code changes from: [b0a0049](https://github.com/openconfig/featureprofiles/commit/b0a00495b8028989dfbd88c72200d31ef6b66f7e)
> * Updates for the new way of referencing deviations.

changes updated

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go [hunk 4]
```diff
@@ -40,7 +40,8 @@ const (
 	ISISInstance = "DEFAULT"
 	// PeerGrpName is BGP peer group name.
 	PeerGrpName = "BGP-PEER-GROUP"
-
+	// PeerGrpEgressName is Egress port BGP peer group name.
+	PeerGrpEgressName = "BGP-PEER-GROUP-EGRESS"
```
Comment (jasdeep-hundal) [inline feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go:44]: Should be removed?

**Review by jasdeep-hundal — CHANGES_REQUESTED** [review]

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go (hunk 4 above)
Comment (ANISH-GOTTAPU) [reply feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go]: yes, should be removed. updated

File: feature/experimental/system/gnmi/benchmarking/otg_tests/drained_configuration_convergence_time/drained_configuration_convergence_time_isis_test.go
Comment (jasdeep-hundal) [inline feature/experimental/system/gnmi/benchmarking/otg_tests/drained_configuration_convergence_time/drained_configuration_convergence_time_isis_test.go:1]: Please incorporate the deviation added in https://github.com/openconfig/featureprofiles/commit/5c59e1b4b39a1f12dd09bc7c79aeef549e6ec005#diff-0194658194a1232d72417ad6dce8959ab123e5c6e380c0468e5c27f2ee21b0aa

**Review by jasdeep-hundal — CHANGES_REQUESTED** [review]

File: feature/experimental/system/gnmi/benchmarking/otg_tests/drained_configuration_convergence_time/metadata.textproto
[the remark this answers is not in the document]
Comment (octpetre) [reply feature/experimental/system/gnmi/benchmarking/otg_tests/drained_configuration_convergence_time/metadata.textproto:1]: Done

File: feature/experimental/system/gnmi/benchmarking/otg_tests/internal/setup/setup.go [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+			niIntf.Subinterface = ygot.Uint32(0)
+			niIntf.Id = ygot.String(dp.Name() + ".0")
+		}
+	}
+	return d
+}
+
+// ConfigureATE function is to configure ate ports with ipv4 , bgp
+// and isis peers.
+func ConfigureATE(t *testing.T, ate *ondatra.ATEDevice) {
```
[the remark this answers is not in the document]
Comment (octpetre) [reply feature/experimental/system/gnmi/benchmarking/otg_tests/internal/setup/setup.go]: Updated

File: feature/experimental/system/gnmi/benchmarking/otg_tests/internal/setup/setup.go [hunk 6]
```diff
@@ -0,0 +1,403 @@
...
+		eth.Connection().SetPortName(dp.ID())
+		mac := fmt.Sprintf("02:00:01:01:01:%02x", byte(i&0xff))
+
+		eth.SetMac(mac)
+
+		ip := eth.Ipv4Addresses().Add().SetName(dev.Name() + ".IPv4")
+		ip.SetAddress(ATEIPList[dp.ID()].String()).SetGateway(DUTIPList[dp.ID()].String()).SetPrefix(int32(plenIPv4))
+
+		// Add BGP routes and ISIS routes , ate port1 is ingress port.
+		if dp.ID() == "port1" {
```
[the remark this answers is not in the document]
Comment (octpetre) [reply feature/experimental/system/gnmi/benchmarking/otg_tests/internal/setup/setup.go]: Done

File: feature/experimental/system/gnmi/benchmarking/otg_tests/drained_configuration_convergence_time/drained_configuration_convergence_time_bgp_test.go [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+// TestEstablish is to configure Interface, BGP and ISIS configurations on DUT
+// using gnmi set request. It also verifies for bgp and isis adjacencies.
+func TestEstablish(t *testing.T) {
+
+	dut := ondatra.DUT(t, "dut")
+	dutConfigPath := gnmi.OC()
+
+	t.Log("Configure Network Instance type to DEFAULT on DUT.")
+	dutConfNIPath := gnmi.OC().NetworkInstance(deviations.DefaultNetworkInstance(dut))
+	gnmi.Replace(t, dut, dutConfNIPath.Type().Config(), oc.NetworkInstanceTypes_NETWORK_INSTANCE_TYPE_DEFAULT_INSTANCE)
```
[the remark this answers is not in the document]
Comment (octpetre) [reply feature/experimental/system/gnmi/benchmarking/otg_tests/drained_configuration_convergence_time/drained_configuration_convergence_time_bgp_test.go]: Done

File: feature/experimental/system/gnmi/benchmarking/otg_tests/drained_configuration_convergence_time/drained_configuration_convergence_time_bgp_test.go [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+			// 	return present
+			// }).Await(t)
+
+			// _, ok := gnmi.WatchAll(t, ate.OTG(), prefixPath.Med().State(), 5*time.Minute, func(v *ygnmi.Value[otgtelemetry.E_UnicastIpv4Prefix_Origin]) bool {
+			// 	gotSetMED, present := v.Val()
+			// 	return present && cmp.Diff(wantSetMED, gotSetMED) == ""
+			// }).Await(t)
+			// if !ok {
+			// 	t.Errorf("obtained MED on ATE is not as expected")
+			// }
```
[the remark this answers is not in the document]
Comment (octpetre) [reply feature/experimental/system/gnmi/benchmarking/otg_tests/drained_configuration_convergence_time/drained_configuration_convergence_time_bgp_test.go]: Enabled

File: feature/experimental/system/gnmi/benchmarking/otg_tests/drained_configuration_convergence_time/drained_configuration_convergence_time_bgp_test.go
[the remark this answers is not in the document]
Comment (octpetre) [reply feature/experimental/system/gnmi/benchmarking/otg_tests/drained_configuration_convergence_time/drained_configuration_convergence_time_bgp_test.go:1]: Done

File: feature/experimental/system/gnmi/benchmarking/otg_tests/drained_configuration_convergence_time/drained_configuration_convergence_time_isis_test.go
Comment (octpetre) [reply feature/experimental/system/gnmi/benchmarking/otg_tests/drained_configuration_convergence_time/drained_configuration_convergence_time_isis_test.go:1]: Done

File: feature/experimental/system/gnmi/benchmarking/otg_tests/drained_configuration_convergence_time/metadata.textproto
[the remark this answers is not in the document]
Comment (jasdeep-hundal) [reply feature/experimental/system/gnmi/benchmarking/otg_tests/drained_configuration_convergence_time/metadata.textproto:1]: @octpetre: Not seeing these as exactly synced yet.

File: feature/experimental/system/gnmi/benchmarking/otg_tests/internal/setup/setup.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (jasdeep-hundal) [reply feature/experimental/system/gnmi/benchmarking/otg_tests/internal/setup/setup.go]: @octpetre : I don't think I see the updates reflected in the latest code in this patch.

File: feature/experimental/system/gnmi/benchmarking/otg_tests/drained_configuration_convergence_time/drained_configuration_convergence_time_bgp_test.go
[the remark this answers is not in the document]
Comment (jasdeep-hundal) [reply feature/experimental/system/gnmi/benchmarking/otg_tests/drained_configuration_convergence_time/drained_configuration_convergence_time_bgp_test.go:1]: @octpetre : I am not seeing the changes from the following in the latest patch:
- https://github.com/openconfig/featureprofiles/pull/2459/files
- https://github.com/openconfig/featureprofiles/commit/5c59e1b4b39a1f12dd09bc7c79aeef549e6ec005#diff-0194658194a1232d72417ad6dce8959ab123e5c6e380c0468e5c27f2ee21b0aaR51-R59

File: feature/experimental/system/gnmi/benchmarking/otg_tests/drained_configuration_convergence_time/drained_configuration_convergence_time_isis_test.go
Comment (jasdeep-hundal) [reply feature/experimental/system/gnmi/benchmarking/otg_tests/drained_configuration_convergence_time/drained_configuration_convergence_time_isis_test.go:1]: @octpetre : Not yet seeing this in the latest patch.

**Review by jasdeep-hundal — CHANGES_REQUESTED** [review]

File: feature/experimental/system/gnmi/benchmarking/otg_tests/drained_configuration_convergence_time/metadata.textproto [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+  platform: {
+    vendor: NOKIA
+      }
+  deviations: {
+    isis_global_authentication_not_required: true
+    isis_explicit_level_authentication_config: true
+    missing_isis_interface_afi_safi_enable: true
+    explicit_port_speed: true
+    explicit_interface_in_default_vrf: true
+interface_enabled: true
```
[the remark this answers is not in the document]
Comment (octpetre) [reply feature/experimental/system/gnmi/benchmarking/otg_tests/drained_configuration_convergence_time/metadata.textproto:18]: corrected

File: feature/experimental/system/gnmi/benchmarking/otg_tests/drained_configuration_convergence_time/drained_configuration_convergence_time_bgp_test.go [hunk 10]
```diff
@@ -0,0 +1,333 @@
...
+	} else {
+		pgpolicy := afipg.GetOrCreateApplyPolicy()
+		pgpolicy.ImportPolicy = policy
+		gnmi.Replace(t, dut, gnmi.OC().NetworkInstance(deviations.DefaultNetworkInstance(dut)).Protocol(oc.PolicyTypes_INSTALL_PROTOCOL_TYPE_BGP, "BGP").Bgp().PeerGroup(setup.PeerGrpName).Config(), pg)
+
+	}
+}
+
+// isConverged function is used to check if ATE has received all the prefixes.
+func isConverged(t *testing.T, dut *ondatra.DUTDevice, ate *ondatra.ATEDevice, ap *ondatra.Port) {
```
[the remark this answers is not in the document]
Comment (octpetre) [reply feature/experimental/system/gnmi/benchmarking/otg_tests/drained_configuration_convergence_time/drained_configuration_convergence_time_bgp_test.go:137]: delay added

File: feature/experimental/system/gnmi/benchmarking/internal/setup/setup.go [hunk 11]
```diff
@@ -329,6 +330,119 @@ func ConfigureATE(t *testing.T, ate *ondatra.ATEDevice) {
...
 }
 
+// ConfigureOTG function is to configure ate ports with ipv4 , bgp
+// and isis peers.
+
+func ConfigureOTG(t *testing.T, ate *ondatra.ATEDevice) {
+	otg := ate.OTG()
+	topo := gosnappi.NewConfig()
+
+	for i, dp := range ate.Ports() {
```
[the remark this answers is not in the document]
Comment (octpetre) [reply feature/experimental/system/gnmi/benchmarking/internal/setup/setup.go:340]: done

File: feature/experimental/system/gnmi/benchmarking/internal/setup/setup.go [hunk 12]
```diff
@@ -329,6 +330,71 @@ func ConfigureATE(t *testing.T, ate *ondatra.ATEDevice) {
 	topo.StartProtocols(t)
 }
 
+// ConfigureOTG function is to configure ate ports with ipv4 , bgp
+// and isis peers.
+
```
Comment (jasdeep-hundal) [inline feature/experimental/system/gnmi/benchmarking/internal/setup/setup.go:335]: Remove the space here (and the comma after `ipv4`).

**Review by jasdeep-hundal — CHANGES_REQUESTED** [review]

File: feature/experimental/system/gnmi/benchmarking/internal/setup/setup.go (hunk 12 above)
Comment (octpetre) [reply feature/experimental/system/gnmi/benchmarking/internal/setup/setup.go:335]: probably this new commented line is better

**Review by jasdeep-hundal — APPROVED** [review]

**Pull request merged**